File: examples/ch04/ts/notify.ts
// Notification — answers to one actor: marketing (the receipt message). One reason
// to change: what the customer is told, and how. The email I/O is isolated here,
// away from the money math.

import { appendFileSync } from "node:fs";
import { type Order } from "./models";

// Pure: the message text. Marketing owns this; no I/O to test around.
export const receiptBody = (order: Order, total: number): string =>
  `Thanks ${order.customer.name}! You paid ${total.toFixed(2)}.`;

// The I/O edge. The "send" is faked as a local file append so the example runs with
// no SMTP server; failures never block a sale.
export const sendReceipt = (
  order: Order,
  total: number,
  outboxFile = "outbox.txt",
): void => {
  try {
    appendFileSync(outboxFile, `To: ${order.customer.email}\n${receiptBody(order, total)}\n`);
  } catch {
    // a failed email must never block a sale
  }
};

File: examples/ch04/notify.py
"""Notification — answers to one actor: marketing (the receipt message).

One reason to change: what the customer is told, and how. The email I/O is
isolated here, away from the money math.
"""

import smtplib
from email.message import EmailMessage

from models import Order


def receipt_body(order: Order, total: float) -> str:
    """Pure: the message text. Marketing owns this; no I/O to test around."""
    return f"Thanks {order.customer.name}! You paid {total:.2f}."


def send_receipt(order: Order, total: float) -> None:
    """The I/O edge. Failures never block a sale."""
    msg = EmailMessage()
    msg["Subject"] = "Your checkout-lite receipt"
    msg["To"] = order.customer.email
    msg.set_content(receipt_body(order, total))
    try:
        with smtplib.SMTP("localhost", 25, timeout=1) as smtp:
            smtp.send_message(msg)
    except OSError:
        pass
